core/chains/evm/logpoller/log_poller.go

```diff
@@ -76,24 +86,92 @@ func NewLogPoller(orm *ORM, ec client.Client, lggr logger.Logger, pollPeriod tim
 		pollPeriod:        pollPeriod,
 		finalityDepth:     finalityDepth,
 		backfillBatchSize: backfillBatchSize,
-		addresses:         make(map[common.Address]struct{}),
-		eventSigs:         make(map[common.Hash]struct{}),
+		events:            make(map[EventID]struct{}),
+		callbacks:         make(map[EventID]Callback),
 	}
 }
 
-// MergeFilter adds the provided event signatures and addresses to the log poller's log filter query.
+func (lp *logPoller) ChainID() *big.Int {
+	return lp.orm.chainID
+}
+
+// MergeFilter adds the provided eventID to the log poller's log filter query.
 // If an event matching any of the given event signatures is emitted from any of the provided addresses,
 // the log poller will pick those up and save them. For topic specific queries see content based querying.
 // Clients may choose to MergeFilter and then replay in order to ensure desired logs are present.
-func (lp *logPoller) MergeFilter(eventSigs []common.Hash, addresses []common.Address) {
+// NOTE: due to constraints of the eth filter, there is "leakage" between successive MergeFilter calls, for example
+// MergeFilter([]EventID{{event1, addr1}}
+// MergeFilter([]EventID{{event2, addr2}}
+// will result in the poller saving (event1, addr2) or (event2, addr1) as well, should it exist.
+// Generally speaking this is harmless. We enforce that the filter contains an address and eventID,
+// which means that anonymous events are not supported and log.Topics >= 1 always (log.Topics[0] is the event signature).
+func (lp *logPoller) MergeFilter(eventIDs []EventID) error {
+	lp.filterMu.Lock()
+	defer lp.filterMu.Unlock()
+	for _, eventID := range eventIDs {
+		// Force specification of both address and eventSig to avoid
+		// events from unknown addresses.
+		if eventID.EventSig == [common.HashLength]byte{} {
+			return errors.Errorf("empty event sig")
+		}
+		if eventID.Address == [common.AddressLength]byte{} {
+			return errors.Errorf("empty address")
+		}
+		lp.events[eventID] = struct{}{}
+	}
+	return nil
+}
+
+type EventID struct {
+	EventSig common.Hash
+	Address  common.Address
+}
+
+type Callback struct {
+	Save  func(tx pg.Queryer, logs []types.Log) error
+	Reorg func(tx pg.Queryer, start, end int64) error
```

Comment:
Do we need to explicitly pass a context to any of these or is there a context implicitly contained in the `pg.Queryer` object? I imagine these need to run in a reasonable amount of time right?

Reply:
there's a context inside, see https://github.com/smartcontractkit/chainlink/blob/5d2a3ce05a73102859a9cbe746f1fc9958462eca/core/services/pg/q.go#L140 where it'll use the DefaultQueryTimeout